### ISSUE TITLE
API Errors return 404

### ISSUE THREAD

**Issue by slifty** [PCNI/OpenHMIS#19]
Right now our API either returns a valid object, or returns a 404.  Ideally the API would return an appropriate error message informing the API user what went wrong.  For example, if the API user attempted to access an entity that she does not have access to, the API should return a message stating as much.

**Comment by ashaar364** [comment]
Api is updated to return the error message if something went wrong, it give the errorcode, application message and developer message.

**Comment by cecilia-donnelly** [comment]
@ashaar364 @slifty:

It looks like these changes were made in commit e25c7cb, in the OAuth2.0.1 branch.  Is that right, @ashaar364?  Can/should they go into master?  Or perhaps this should be our first "feature" branch as described in the new CONTRIBUTING.md.  Thoughts, @slifty?

**Comment by ashaar364** [comment]
@cecilia-donnelly That is right the changes were made in the branch OAuth2.0.1 and yes they can go into the master.

**Comment by slifty** [comment]
That is awesome @ashaar364 !  Would you be around at all some time this week to talk through the proposed contribution process stuff that _just_ got added to the repository?  This would be an excellent chance to try it out in the wild.

(Happy to talk through it by email too if time is tight, although it would be good to meet!)

**Comment by ashaar364** [comment]
@slifty we can meet Friday at 5pm ur time. Let me know if that's fine with you. By that time let me have a look at the contribution process.

**Comment by slifty** [comment]
@ashaar364 perfect; lets use Skype for now, I'm "slifty" there too.

**Comment by kfogel** [comment]
Be careful with information leakage in errors.  For example, returning a "you don't have authorization to access ${ENTITY}" error confirms to the API user that ${ENTITY} exists, which itself may be a privacy violation under some quite-easy-to-imagine scenarios.

Totally agree about returning comprehensible errors wherever possible.  We just need to make sure we don't leak inappropriate information in those errors.

**Comment by ashaar364** [comment]
code is committed in the features-errors branch
